src/particle.cpp

```diff
@@ -406,9 +408,15 @@ void Particle::event_revive_from_secondary()
       // have to determine it before the energy of the secondary particle can be
       // removed from the pulse-height of this cell.
       if (lowest_coord().cell == C_NONE) {
-        if (!exhaustive_find_cell(*this)) {
-          mark_as_lost("Could not find the cell containing particle " +
-                       std::to_string(id()));
+        bool verbose = settings::verbosity >= 10 || trace();
+        if (!exhaustive_find_cell(*this, verbose)) {
+          // Note: Geometron:: prefix is necessary here because it's ambiguous
+          // whether Particle::mark_as_lost(const char*) should be called with
+          // implicit argument conversion or Geometron::mark_as_lost(const
+          // std::string&) should be called.
+          Geometron::mark_as_lost(
+            "Could not find the cell containing particle " +
+            std::to_string(id()));
```

Comment:
If a normal MC particle is lost, would calling the `Geometron::mark_as_lost()` version of the function result in problems as e.g. the particle weight doesn't get set to zero? Or that a lost particle results in a fatal error instead of just a warning message?

Reply:
No, so, this is the non-polymorphic call. It's a little confusing. Since this is a passthrough to the virtual `mark_as_lost`, the correct behavior comes here.

I'm not sure how to express this otherwise. Open to suggestions! I can't get it to compile by just using `mark_as_lost` alone because it can't determine if there is an implicit cast happening on that argument.

Reply:
I'm not sure I follow -- I would think that the goal is to use the derived type's `mark_as_lost()` function? If so, why explicitly call the base class version? As you said, just using `mark_as_lost()` (without the scope resolution operator) should result in the correct polymorphic evaluation, where the derived function gets used as intended. As for resolving the overload issue, can the other `mark_as_lost()`'s be made virtual as well (along with having to define them again in `Particle`)? As it is now, it looks like only one of the three is virtual? Maybe there is an easier way though! If I'm not seeing this right though -- let me know! Just want to be sure we end up using the derived class `mark_as_lost()` function, as we don't want to error out due to one lost particle in a large sim.

Reply:
It does call the derived type, by calling the base class's non-virtual that dispatches the virtual. I know it ain't pretty.

Hm, I'll try the approach of making the other two virtual as well. I think that will let it compile?

Reply:
So, here's the problem with not being explicit about calling the base class version. It will try to call the virtual specialization that `Particle` owns and convert the argument to `const char*`, but there is no way to do that. For some reason, it doesn't try to call the base class's version unless I'm explicit about it. I don't know why that's the case--must be some detailed precedence rules in the standard.

```
[  1%] Building CXX object CMakeFiles/libopenmc.dir/src/particle.cpp.o
/home/gavin/Code/openmc/src/particle.cpp: In member function ‘void openmc::Particle::event_calculate_xs()’:
/home/gavin/Code/openmc/src/particle.cpp:163:56: error: cannot convert ‘std::__cxx11::basic_string<char>’ to ‘const char*’
  163 |         "Could not find the cell containing particle " + std::to_string(id()));
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
      |                                                        |
      |                                                        std::__cxx11::basic_string<char>
In file included from /home/gavin/Code/openmc/src/particle.cpp:1:
/home/gavin/Code/openmc/include/openmc/particle.h:105:41: note:   initializing argument 1 of ‘virtual void openmc::Particle::mark_as_lost(const char*)’
  105 |   virtual void mark_as_lost(const char* message) override;
      |                             ~~~~~~~~~~~~^~~~~~~

```

Reply:
Just checked... adding the virtual keyword to the other two doesn't help. Ugh.

Reply:
Ohhh. Just figured this out. Overloads get hidden in derived classes. Have to use a `using` statement to clarify it. C++ has no end to its depth!

https://stackoverflow.com/a/8816835

Reply:
Alright, got it fixed in the new commit.

Reply:
This looks good now! I think the issue originally was that if calling as `Geometron::mark_as_lost("Error!")` it would have been hitting the `fatal_error()` version rather than the dispatching to virtual versions. With the way you've changed it, it should now call the intended derived types regardless of what sort of `string` vs. `char*` etc is passed, so looks to be in good shape! Thanks for making the change.